src/main/java/org/apache/sysml/hops/ipa/InterProceduralAnalysis.java

```diff
@@ -543,7 +546,27 @@ else if (sb instanceof ForStatementBlock) //incl parfor
 			propagateStatisticsIntoFunctions(prog, roots, fcand, callVars, fcandSafeNNZ, unaryFcands, fnStack);
 		}
 	}
-	
+
+	/**
+	 * Propagate scalar values across DAGs.
+	 *
+	 * This replaces scalar transient reads with literals.
+	 *
+	 * @param roots  List of HOPs.
+	 * @param vars  Map of variables eligible for propagation.
+	 * @throws HopsException
+	 */
+	private void propagateScalarsAcrossDAG(ArrayList<Hop> roots, LocalVariableMap vars)
+		throws HopsException
+	{
+		for (Hop hop : roots) {
+			try {
+				Recompiler.rReplaceLiterals(hop, vars);
```

Comment:
We now expose more fine-grained control over literal replacement - please use rReplaceLiterals(hop, vars, true), which restricts literal replacements to scalars only, disallowing replacing small matrix operations as it would fail for temporary matrix object that we use for size propagation.